learn/security/tenant_tokens.md

```diff
@@ -0,0 +1,299 @@
+    'filter': `user_id = ${currentUserID}`
+  }
+};
+const token = client.generateTenantToken(searchRules, expiryDate, apiKey);
+```
+
+::: danger
+Because tenant tokens are generated in your application, Meilisearch cannot check if search rule filters are valid. Invalid search rules will only throw errors when they are used in a query.
+
+Consult the search API reference [more information on Meilisearch filter syntax](/reference/api/search.md#filter).
+:::
+
+### API Keys
+
+Creating a token requires an API key. A token has access to the same indexes as the API key used to generate it.
+
+Since a master key is not an API key, **you cannot use a master key to create a tenant token**.
+
+For security reasons, we strongly recommend you avoid exposing the API key whenever possible and **always generate tokens on your application's back-end**.
+
+When using an official Meilisearch SDK, you may indicate which API key you wish to use when generating a token. Consult the documentation of the SDK you are using for more specific instructions.
+
+In [the previous SDK example](#generating-tenant-tokens-with-an-sdk), the API key is stored in `apiKey`:
+
+```js
+// placeholder code sample. Final version will be supplied by the integrations team
+// <CodeSamples id="tenant_token_guide_fragment_key_1">
+const apiKey = '[an_api_key]';
+const token = client.generateTenantToken(searchRules, expiryDate, apiKey);
+```
+
+::: note
+If an API key expires, any token tenants created with it will become invalid. The same applies if the API key is deleted or regenerated due to a changed master key.
+:::
+
+[You can read more about API keys in our dedicated guide.](/learn/security/master_api_keys.md)
+
+### Expiry date
+
+It is possible to define an expiry date when generating a token. This is good security practice and Meilisearch recommends setting relatively short token expiry dates whenever possible.
+
+The expiry date must be a UNIX timestamp or `null`. Additionally, a token's expiration date cannot exceed its parent API key's expiration date. For example, if an API key is set to expire on 2022-10-15, a token generated with that API key cannot be set to expire on 2022-10-16.
+
+Setting a token expiry date is optional, but recommended. Tokens without an expiry date are never expire and can be used indefinitely as long as its parent API key remains valid.
+
+::: danger
+The only way to revoke a token without an expiry date is to [delete](/reference/api/keys.md#delete-a-key) its parent API key.
+
+Changing an instance's master key forces Meilisearch to regenerate all API keys and will also render all existing tenant tokens invalid.
+:::
+
+When using an official Meilisearch SDK, you may indicate the expiry date when generating a token. Consult the documentation of the SDK you are using for more specific instructions.
+
+In [the previous SDK example](#generating-tenant-tokens-with-an-sdk), the expiry date is stored in `expiryDate`:
+
+```js
```

Comment:
Not sure where to put this comment, but I am wondering if these three "code sample fragments" are really helpful. I think I understand the value of having a visual/code aid in this reference section, but aren't the payload elements/method arguments being discussed pretty SDK-agnostic? I imagine that it would be enough to show an example value for each payload element in a simple triple-backtick code block.

Then again, maybe there's some value I'm not considering, such as making it easy for users of typed languages to see the recommended data type for each argument. But even then, it's nothing that wouldn't be included in "tenant_token_guide_generate_sdk_1" above. So I'm inclined to replace these with inline code blocks rather than trouble the integrations team unnecessarily. WDYT?

Reply:
I understand what you mean, but it felt more user-friendly to have the SDK options here too. Especially in the case a user has already chosen an SDK (and thus explicitly named the language in which they want to see code samples), so it can be annoying/make understanding the sample a tiny bit harder if I'm e.g. a Rust dev and suddenly see some JS.

Also, and this might be more important, different SDKs have slightly different signatures (I believe JS accepts one search rules object and another for `exp` and `apiKey`, whilst Ruby accepts three separate arguments for the same values).